src/nat/middleware/dynamic/dynamic_function_middleware.py

```
@@ -0,0 +1,1069 @@
+        for function_name in all_functions:
+            try:
+                self._register_component_function(discovered_component, function_name)
+            except Exception as e:
+                logger.debug("Failed to register component function '%s' on auth provider '%s': %s",
+                             function_name,
+                             auth_provider_name,
+                             e)
+
+        return auth_client
+
+    async def _discover_and_register_function(self, name: str | FunctionRef) -> Function:
```

Ruff (0.14.7)

Undefined name `FunctionRef`

(F821)